src/transforms/pipelines/config.rs

```diff
@@ -41,62 +42,115 @@ impl Clone for PipelineConfig {
+                Transform::Synchronous(transform) => transform,
+                _ => panic!("non-sync transform in pipeline: {:?}", config),
+            };
+            transforms.push(transform);
+        }
+        Ok(Transform::Synchronous(Box::new(Pipeline {
+            condition,
+            transforms,
+        })))
+    }
+
+    fn input(&self) -> Input {
+        Input::all()
+    }
+
+    fn outputs(&self, _: &schema::Definition) -> Vec<Output> {
+        vec![Output::default(DataType::all())]
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "pipeline"
+    }
+
+    fn enable_concurrency(&self) -> bool {
+        true
+    }
+}
+
 impl PipelineConfig {
     pub(super) fn expand(
         &mut self,
         name: &ComponentKey,
         inputs: &[String],
     ) -> crate::Result<Option<InnerTopology>> {
         let mut result = InnerTopology::default();
-        // define the name of the last output
-        let last_name = if self.transforms.is_empty() {
-            self.filter
-                .as_ref()
-                .map(|_filter| {
-                    let filter_name = name.join("filter");
-                    filter_name.join("success")
-                })
-                .ok_or_else(|| "mut have at least one transform or a filter".to_string())?
-        } else {
-            name.join(self.transforms.len() - 1)
-        };
+
+        result.inner.insert(
+            name.clone(),
+            InnerTopologyTransform {
+                inputs: inputs.to_vec(),
+                inner: Box::new(self.clone()),
+            },
+        );
+        // TODO: actually call outputs fn
         result
             .outputs
-            .push((last_name, vec![Output::default(DataType::all())]));
-        // insert the filter if needed and return the next inputs
-        let mut next_inputs = if let Some(ref filter) = self.filter {
-            let mut conditions = IndexMap::new();
-            conditions.insert("success".to_string(), filter.to_owned());
-            let filter_name = name.join("filter");
-            result.inner.insert(
-                filter_name.clone(),
-                InnerTopologyTransform {
-                    inputs: inputs.to_vec(),
-                    inner: Box::new(RouteConfig::new(conditions)),
-                },
-            );
-            result.outputs.push((
-                filter_name.clone(),
-                vec![Output::default(DataType::all()).with_port(UNMATCHED_ROUTE)],
-            ));
-            vec![filter_name.port("success")]
+            .push((name.clone(), vec![Output::default(DataType::all())]));
+
+        Ok(Some(result))
+    }
+}
+
+#[derive(Clone)]
+struct Pipeline {
+    condition: Option<Condition>,
+    transforms: Vec<Box<dyn SyncTransform>>,
+}
+
+impl SyncTransform for Pipeline {
+    fn transform(&mut self, _event: Event, _output: &mut TransformOutputsBuf) {
+        unimplemented!()
```

Comment:
This is a bit of a wart in the `SyncTransform` API. We could consider splitting out another `BatchTransform` variant for `transform_all` and implementing the trait for `SyncTransform` instead.